wolfcrypt/src/wc_lms.c

```diff
@@ -19,8 +19,1139 @@
+ */
+#define LMS_PARAMS(l, h, w, wb, t, t2)                              \
+    { l, h, w, LMS_LS(w, wb), LMS_P(w, wb), t, t2,                  \
+      LMS_PARAMS_SIG_LEN(l, h, LMS_P(w, wb)), LMS_PARAMS_CACHE(h) }
+
+
+/* Initialize the working state for LMS operations.
+ *
+ * @param [in, out] state   LMS state.
+ * @param [in]      params  LMS parameters.
+ */
+static int wc_lmskey_state_init(LmsState* state, const LmsParams* params)
+{
+    int ret;
 
-#ifdef WOLFSSL_HAVE_LMS
-    #error "Contact wolfSSL to get the implementation of this file"
+    /* Zero out every field. */
+    XMEMSET(state, 0, sizeof(LmsState));
+
+    /* Keep a reference to the parameters for use in operations. */
+    state->params = params;
+
+    /* Initialize the two hash algorithms. */
+    ret = wc_InitSha256(&state->hash);
+    if (ret == 0) {
+        ret = wc_InitSha256(&state->hash_k);
+        if (ret != 0) {
+            wc_Sha256Free(&state->hash);
+        }
+    }
+
+    return ret;
+}
+
+/* Free the working state for LMS operations.
+ *
+ * @param [in] state  LMS state.
+ */
+static void wc_lmskey_state_free(LmsState* state)
+{
+    wc_Sha256Free(&state->hash_k);
+    wc_Sha256Free(&state->hash);
+}
+
+/* Supported LMS parameters. */
+static const wc_LmsParamsMap wc_lms_map[] = {
+#if LMS_MAX_HEIGHT >= 15
+    { WC_LMS_PARM_NONE     , "LMS_NONE"         ,
+      LMS_PARAMS(1, 15, 2, 1, LMS_SHA256_M32_H15, LMOTS_SHA256_N32_W2) },
+    { WC_LMS_PARM_L1_H15_W2, "LMS/HSS L1_H15_W2",
+      LMS_PARAMS(1, 15, 2, 1, LMS_SHA256_M32_H15, LMOTS_SHA256_N32_W2) },
+    { WC_LMS_PARM_L1_H15_W4, "LMS/HSS L1_H15_W4",
+      LMS_PARAMS(1, 15, 4, 2, LMS_SHA256_M32_H15, LMOTS_SHA256_N32_W4) },
+#endif
+#if LMS_MAX_LEVELS >= 2
+#if LMS_MAX_HEIGHT >= 10
+    { WC_LMS_PARM_L2_H10_W2, "LMS/HSS L2_H10_W2",
+      LMS_PARAMS(2, 10, 2, 1, LMS_SHA256_M32_H10, LMOTS_SHA256_N32_W2) },
+    { WC_LMS_PARM_L2_H10_W4, "LMS/HSS L2_H10_W4",
+      LMS_PARAMS(2, 10, 4, 2, LMS_SHA256_M32_H10, LMOTS_SHA256_N32_W4) },
+    { WC_LMS_PARM_L2_H10_W8, "LMS/HSS L2_H10_W8",
+      LMS_PARAMS(2, 10, 8, 3, LMS_SHA256_M32_H10, LMOTS_SHA256_N32_W8) },
+#endif
+#endif
+#if LMS_MAX_LEVELS >= 3
+    { WC_LMS_PARM_L3_H5_W2 , "LMS/HSS L3_H5_W2" ,
+      LMS_PARAMS(3,  5, 2, 1, LMS_SHA256_M32_H5 , LMOTS_SHA256_N32_W2) },
+    { WC_LMS_PARM_L3_H5_W4 , "LMS/HSS L3_H5_W4" ,
+      LMS_PARAMS(3,  5, 4, 2, LMS_SHA256_M32_H5 , LMOTS_SHA256_N32_W4) },
+    { WC_LMS_PARM_L3_H5_W8 , "LMS/HSS L3_H5_W8" ,
+      LMS_PARAMS(3,  5, 8, 3, LMS_SHA256_M32_H5 , LMOTS_SHA256_N32_W8) },
+#if LMS_MAX_HEIGHT >= 10
+    { WC_LMS_PARM_L3_H10_W4, "LMS/HSS L3_H10_W4",
+      LMS_PARAMS(3, 10, 4, 2, LMS_SHA256_M32_H10, LMOTS_SHA256_N32_W4) },
+#endif
+#endif
+#if LMS_MAX_LEVELS >= 4
+    { WC_LMS_PARM_L4_H5_W8 , "LMS/HSS L4_H5_W8" ,
+      LMS_PARAMS(4,  5, 8, 3, LMS_SHA256_M32_H5 , LMOTS_SHA256_N32_W8) },
+#endif
+
+    /* For when user sets L, H, W explicitly. */
+    { 0x100                , "LMS/HSS_L1_H5_W1" ,
```

Comment:
This part also out dated and not using updated enum.

Reply:
fixed